src/Microsoft.AspNet.Mvc.ModelBinding/Internal/DictionaryHelper.cs

```diff
@@ -28,6 +29,16 @@ public static class DictionaryHelper
                     continue;
                 }
 
+                if (key.StartsWith("[", StringComparison.OrdinalIgnoreCase))
```

Comment:
Yes, this is needed for a case where model is a collection like `List<Product>`. In this case MSD has entries like `[0].ProductName`, `[1].ProductName`.This is scenario not specific to clearing model state. Even for detemining ValidationState in a collections, we will have to do the same right? I ran modelbinding unit tests and functional tests and they passed.